app/src/main/java/com/example/weather/MainActivity.java

```diff
@@ -106,8 +121,34 @@ public void run() {
     public void onClickByLocation(View view) {
         // todo: implement
         // https://trello.com/c/W4VxNHog
+        Log.d(TAG, "onClickByLocation start");
+
+        locationClient.getLocation(onLocationCallback);
+
+        Log.d(TAG, "onClickByLocation finish");
     }
 
+    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
+        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
+
+        switch (requestCode) {
+            case LocationClient.LOCATION_REQUEST: {
+                if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
+                    locationClient.getLocation(onLocationCallback);
+                } else if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_DENIED) {
+                    Toast.makeText(this, "Permission denied", Toast.LENGTH_SHORT).show();
+                } else {
+                    Toast.makeText(this, "Permission denied", Toast.LENGTH_SHORT).show();
+                }
+                break;
+            }
+        }
+    }
```

Comment:
move out this logic into LocationClient
```suggestion
locationClient.onRequestPermissionsResult(requestCode, permissions, grandResults);
```